sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,473 @@
+
+      source.populateDisplayData(builder);
+    }
+
+  }
+
+  private static class UnboundedMqttSource extends UnboundedSource<byte[], MqttCheckpointMark> {
+
+    public UnboundedMqttSource withServerUri(String serverUri) {
+      return new UnboundedMqttSource(serverUri, clientId, topic, maxNumRecords, maxReadTime);
+    }
+
+    public UnboundedMqttSource withClientId(String clientId) {
+      return new UnboundedMqttSource(serverUri, clientId, topic, maxNumRecords, maxReadTime);
+    }
+
+    public UnboundedMqttSource withTopic(String topic) {
+      return new UnboundedMqttSource(serverUri, clientId, topic, maxNumRecords, maxReadTime);
+    }
+
+    public UnboundedMqttSource withMaxNumRecords(long maxNumRecords) {
+      return new UnboundedMqttSource(serverUri, clientId, topic, maxNumRecords, maxReadTime);
+    }
+
+    public UnboundedMqttSource withMaxReadTime(Duration maxReadTime) {
+      return new UnboundedMqttSource(serverUri, clientId, topic, maxNumRecords, maxReadTime);
+    }
+
+    private final String serverUri;
+    private final String clientId;
+    @Nullable
+    private final String topic;
+
+    private final long maxNumRecords;
+    private final Duration maxReadTime;
+
+    public UnboundedMqttSource(String serverUri, String clientId, String topic,
+                               long maxNumRecords, Duration maxReadTime) {
+      this.serverUri = serverUri;
+      this.clientId = clientId;
+      this.topic = topic;
+      this.maxNumRecords = maxNumRecords;
+      this.maxReadTime = maxReadTime;
+    }
+
+    @Override
+    public UnboundedReader createReader(PipelineOptions options,
+                                        MqttCheckpointMark checkpointMark) {
+      return new UnboundedMqttReader(this, checkpointMark);
+    }
+
+    @Override
+    public List<UnboundedMqttSource> generateInitialSplits(int desiredNumSplits,
+                                                           PipelineOptions options) {
+      List<UnboundedMqttSource> sources = new ArrayList<>();
+      for (int i = 0; i < desiredNumSplits; i++) {
+        // NB: it's important that user understand the impact of MQTT message QoS
+        sources.add(new UnboundedMqttSource(serverUri, clientId + "-" + i, topic, maxNumRecords,
+            maxReadTime));
+      }
+      return sources;
+    }
+
+    @Override
+    public void validate() {
+      Preconditions.checkNotNull(serverUri, "serverUri");
+      Preconditions.checkNotNull(clientId, "clientId");
+    }
+
+    @Override
+    public Coder getCheckpointMarkCoder() {
+      return SerializableCoder.of(MqttCheckpointMark.class);
+    }
+
+    @Override
+    public Coder getDefaultOutputCoder() {
+      return SerializableCoder.of(byte[].class);
+    }
+
+  }
+
+  private static class UnboundedMqttReader extends UnboundedSource.UnboundedReader
+      implements MqttCallback {
+
+    private UnboundedMqttSource source;
+
+    private MqttCheckpointMark checkpointMark;
```

Comment:
This field is effectively unused; the checkpoint mark is passed around but it doesn't do anything, in particular, it does not affect what data is being read; I suppose because the MQTT protocol actually doesn't support checkpointing and resuming. To better express that, why not just use Void as the checkpoint type, and get rid of MqttCheckpointMark.